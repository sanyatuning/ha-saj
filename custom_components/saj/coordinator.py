"""Coordinator to fetch data from SAJ solar inverter."""
import asyncio
import logging
from urllib.parse import urlparse

from homeassistant.const import (
    CONF_HOST,
    CONF_NAME,
    CONF_PASSWORD,
    CONF_TYPE,
    CONF_USERNAME,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.typing import ConfigType
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from . import pysaj
from .const import DEFAULT_NAME, DOMAIN, INVERTER_TYPES, UPDATE_INTERVAL

_LOGGER = logging.getLogger(__name__)


def _init_pysaj(wifi, host, username, password):  # pragma: no cover
    kwargs = {"wifi": wifi}
    if username and password:
        kwargs["username"] = username
        kwargs["password"] = password

    return pysaj.SAJ(host, **kwargs)


class SAJDataUpdateCoordinator(DataUpdateCoordinator):
    """Representation of a SAJ inverter."""

    def __init__(
        self,
        hass: HomeAssistant,
        config: ConfigType,
    ) -> None:
        """Init SAJ Inverter class."""
        super().__init__(
            hass,
            _LOGGER,
            name=config.get(CONF_NAME) or DEFAULT_NAME,
            update_interval=UPDATE_INTERVAL,
            update_method=self.update,
        )
        self.last_update_success = False
        wifi = config[CONF_TYPE] == INVERTER_TYPES[1]
        self._saj = _init_pysaj(
            wifi, config[CONF_HOST], config[CONF_USERNAME], config[CONF_PASSWORD]
        )
        self._sensor_def = pysaj.Sensors(wifi)

    def get_enabled_sensors(self):
        """Return enabled sensors."""
        return [s for s in self._sensor_def if s.enabled]

    @property
    def serialnumber(self):
        """Return the serial number of the inverter."""
        return self._saj.serialnumber

    @property
    def device_info(self) -> DeviceInfo:
        """Return device info for this device."""
        url = urlparse(self._saj.url)._replace(path="/").geturl()
        return DeviceInfo(
            identifiers={(DOMAIN, self._saj.serialnumber)},
            manufacturer="SAJ",
            name=self.name,
            configuration_url=str(url),
        )

    async def connect(self):
        """Try to connect to the inverter."""
        try:
            success = await self._saj.read(self._sensor_def)
        except asyncio.TimeoutError as err:
            raise CannotConnect from err
        if not success:
            raise CannotConnect

    async def update(self):
        """Fetch data from Inverter."""
        done = await self._saj.read(self._sensor_def)
        if done:
            return self._sensor_def

        raise UpdateFailed


class CannotConnect(ConfigEntryNotReady):
    """Error to indicate we cannot connect."""
