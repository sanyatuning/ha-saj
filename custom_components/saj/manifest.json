{
  "domain": "saj",
  "name": "SAJ Solar Inverter",
  "codeowners": [
    "@sanyatuning"
  ],
  "config_flow": true,
  "dhcp": [{
    "macaddress": "98F4AB*",
    "hostname": "saj-wifi-*"
  }],
  "documentation": "https://github.com/sanyatuning/ha-saj",
  "iot_class": "local_polling",
  "loggers": [
    "pysaj"
  ],
  "version": "2023.1.1"
}