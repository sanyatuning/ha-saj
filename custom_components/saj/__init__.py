"""The saj component."""
import asyncio

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant

from custom_components.saj.const import DOMAIN
from custom_components.saj.coordinator import SAJDataUpdateCoordinator

PLATFORMS = ["sensor"]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up saj from a config entry."""
    inverter = SAJDataUpdateCoordinator(hass, dict(entry.data))
    await inverter.connect()

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = inverter

    # for component in PLATFORMS:
    #     hass.async_create_task(
    #         hass.config_entries.async_forward_entry_setup(entry, component)
    #     )
    #
    # return True
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = all(
        await asyncio.gather(
            *[
                hass.config_entries.async_forward_entry_unload(entry, component)
                for component in PLATFORMS
            ]
        )
    )
    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
