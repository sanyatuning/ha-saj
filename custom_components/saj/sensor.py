"""SAJ solar inverter interface."""
from __future__ import annotations

import logging
from datetime import date

import voluptuous as vol
from homeassistant.components.sensor import (
    PLATFORM_SCHEMA,
    SensorDeviceClass,
    SensorEntity,
    SensorStateClass,
)
from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
from homeassistant.const import (
    CONF_HOST,
    CONF_NAME,
    CONF_PASSWORD,
    CONF_TYPE,
    CONF_USERNAME,
    ENERGY_KILO_WATT_HOUR,
    MASS_KILOGRAMS,
    POWER_WATT,
    TEMP_CELSIUS,
    TEMP_FAHRENHEIT,
    TIME_HOURS,
)
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from . import pysaj
from .const import DEFAULT_NAME, DOMAIN, INVERTER_TYPES
from .coordinator import SAJDataUpdateCoordinator

_LOGGER = logging.getLogger(__name__)

SAJ_UNIT_MAPPINGS = {
    "": None,
    "h": TIME_HOURS,
    "kg": MASS_KILOGRAMS,
    "kWh": ENERGY_KILO_WATT_HOUR,
    "W": POWER_WATT,
    "Â°C": TEMP_CELSIUS,
}

PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_HOST): cv.string,
        vol.Optional(CONF_NAME): cv.string,
        vol.Optional(CONF_TYPE, default=INVERTER_TYPES[0]): vol.In(INVERTER_TYPES),
        vol.Inclusive(CONF_USERNAME, "credentials"): cv.string,
        vol.Inclusive(CONF_PASSWORD, "credentials"): cv.string,
    }
)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
) -> None:
    """Set up the SAJ sensors."""
    inverter: SAJDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
    await inverter.async_refresh()
    async_add_entities(
        SAJSensor(inverter, sensor) for sensor in inverter.get_enabled_sensors()
    )


async def async_setup_platform(
    hass: HomeAssistant,
    config: ConfigType,
    async_add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
    """Set up the SAJ inverter with yaml."""
    _LOGGER.warning(
        "Loading SAJ Solar inverter integration via yaml is deprecated. "
        "Please remove it from your configuration"
    )
    hass.async_create_task(
        hass.config_entries.flow.async_init(
            DOMAIN,
            context={"source": SOURCE_IMPORT},
            data=config,
        )
    )


class SAJSensor(CoordinatorEntity, SensorEntity):
    """Representation of a SAJ sensor."""

    def __init__(
        self, coordinator: SAJDataUpdateCoordinator, pysaj_sensor: pysaj.Sensor
    ) -> None:
        """Initialize the SAJ sensor."""
        super().__init__(coordinator)
        self._attr_unique_id = f"{coordinator.serialnumber}_{pysaj_sensor.name}"
        self._sensor = pysaj_sensor

        if pysaj_sensor.name in ("current_power", "temperature"):
            self._attr_state_class = SensorStateClass.MEASUREMENT
        if pysaj_sensor.name == "total_yield":
            self._attr_state_class = SensorStateClass.TOTAL
        if pysaj_sensor.name == "today_yield":
            self._attr_state_class = SensorStateClass.TOTAL_INCREASING
        self._attr_device_info = coordinator.device_info

    @property
    def name(self):
        """Return the name of the sensor."""
        if self.coordinator.name != DEFAULT_NAME:
            return f"{self.coordinator.name} {self._sensor.name}"

        return f"SAJ Inverter {self._sensor.name}"

    @property
    def native_value(self):
        """Return the state of the sensor."""
        return self._sensor.value

    @property
    def available(self) -> bool:
        """
        Return True if device is available.

        Note: SAJ inverters are powered by DC via solar panels and thus are
        offline after the sun has set, but we keep some sensor available.
        """
        if self._sensor.per_total_basis or (
            self._sensor.per_day_basis and date.today() == self._sensor.date
        ):
            # filter out invalid zero value
            return self._sensor.value is not None and self._sensor.value > 0

        return super().available

    @property
    def native_unit_of_measurement(self):
        """Return the unit the value is expressed in."""
        return SAJ_UNIT_MAPPINGS[self._sensor.unit]

    @property
    def device_class(self):
        """Return the device class the sensor belongs to."""
        if self.native_unit_of_measurement == POWER_WATT:
            return SensorDeviceClass.POWER
        if self.native_unit_of_measurement == ENERGY_KILO_WATT_HOUR:
            return SensorDeviceClass.ENERGY
        if self.native_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
            return SensorDeviceClass.TEMPERATURE
