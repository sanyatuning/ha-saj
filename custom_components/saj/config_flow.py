"""Config flow for saj integration."""
from __future__ import annotations

import logging
from typing import Any

import voluptuous as vol
from homeassistant import config_entries
from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_TYPE, CONF_USERNAME
from homeassistant.data_entry_flow import FlowError, FlowResult
from homeassistant.helpers.typing import ConfigType

from custom_components.saj import pysaj

from .const import DOMAIN, INVERTER_TYPES
from .coordinator import CannotConnect, SAJDataUpdateCoordinator

_LOGGER = logging.getLogger(__name__)


class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Handle a config flow for SAJ."""

    VERSION = 1

    async def async_step_import(self, import_config: ConfigType) -> FlowResult:
        """Import a config entry from configuration.yaml."""
        for entry in self._async_current_entries(include_ignore=True):
            if import_config[CONF_HOST] == entry.data[CONF_HOST]:
                return self.async_abort(reason="already_configured")
        import_config.setdefault(CONF_TYPE, "ethernet")
        import_config.setdefault(CONF_USERNAME, "")
        import_config.setdefault(CONF_PASSWORD, "")
        return await self.async_step_user(import_config)

    async def async_step_user(self, user_input: dict[str, Any] = None) -> FlowResult:
        """Handle the initial step."""
        errors = {}
        if user_input is not None:
            try:
                inverter = SAJDataUpdateCoordinator(self.hass, user_input)
                await inverter.connect()
                await self.async_set_unique_id(inverter.serialnumber)
                self._abort_if_unique_id_configured()

                return self.async_create_entry(title=inverter.name, data=user_input)
            except pysaj.UnauthorizedException:
                errors["base"] = "invalid_auth"
            except CannotConnect:
                errors["base"] = "cannot_connect"
            except FlowError as error:
                raise error
            except Exception as error:  # pylint: disable=broad-except
                _LOGGER.error("Unexpected exception: %s", error)
                errors["base"] = "unknown"
        else:
            user_input = {}

        return self.async_show_form(
            step_id="user",
            data_schema=vol.Schema(
                {
                    vol.Required(CONF_HOST, default=user_input.get(CONF_HOST, "")): str,
                    vol.Required(CONF_TYPE, default=user_input.get(CONF_TYPE)): vol.In(
                        INVERTER_TYPES
                    ),
                    vol.Optional(
                        CONF_USERNAME,
                        "credentials",
                        default=user_input.get(CONF_USERNAME, ""),
                    ): str,
                    vol.Optional(
                        CONF_PASSWORD,
                        "credentials",
                        default=user_input.get(CONF_PASSWORD, ""),
                    ): str,
                }
            ),
            errors=errors,
        )
