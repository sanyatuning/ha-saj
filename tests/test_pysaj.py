"""Tests for pysaj package."""

import pytest
from aioresponses import aioresponses
from pytest_homeassistant_custom_component.common import load_fixture
from syrupy import SnapshotAssertion

from custom_components.saj.pysaj import SAJ, Sensors

host = "192.168.0.9"
testdata_eth = [
    "status1.xml",
    "status2.xml",
    "status3.xml",
]
testdata_wifi = [
    (0, "15020J2012EN02905"),
    (1, "21030G1628XXYYYYY"),
    (2, "14020G1728EN19744"),
]


@pytest.fixture
def mocked():
    with aioresponses() as m:
        yield m


@pytest.mark.parametrize("name", testdata_eth)
async def test_eth(name, mocked, snapshot: SnapshotAssertion):
    wifi = False
    info_eth = load_fixture("info.xml")
    status_eth = load_fixture(name)
    mocked.get("http://192.168.0.9/equipment_data.xml", body=info_eth)
    mocked.get("http://192.168.0.9/real_time_data.xml", body=status_eth)

    sensors = Sensors(wifi)
    saj = SAJ(host, wifi)
    await saj.read(sensors)

    assert saj.serialnumber == "13020J2020EN09010"
    assert [s.name + ":" + str(s.value) for s in sensors] == snapshot


@pytest.mark.parametrize("index,serial", testdata_wifi)
async def test_wifi(index, serial, mocked, snapshot: SnapshotAssertion):
    wifi = True
    info_csv = load_fixture("info.csv")
    status_csv = load_fixture("status.csv")

    sensors = Sensors(wifi)
    saj = SAJ(host, wifi)

    info_line = info_csv.split("\n")[index]
    status_line = status_csv.split("\n")[index]

    mocked.get("http://admin:admin@192.168.0.9/info.php", body=info_line)
    mocked.get("http://admin:admin@192.168.0.9/status/status.php", body=status_line)
    await saj.read(sensors)

    enabled_count = len([s for s in sensors if s.enabled])
    if serial == "14020G1728EN19744":
        # invalid CSV
        assert enabled_count == 0
    else:
        assert enabled_count == 9
        assert [s.name + ":" + str(s.value) for s in sensors] == snapshot

    assert saj.serialnumber == serial
