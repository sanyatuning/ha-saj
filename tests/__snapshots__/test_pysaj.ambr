# serializer version: 1
# name: test_eth[status1.xml]
  list([
    'current_power:2303',
    'grid_voltage:None',
    'grid_l1:241.1',
    'grid_l2:None',
    'grid_l3:None',
    'today_yield:12.55',
    'total_yield:586.85',
    'today_time:6.8',
    'total_time:1278.7',
    'total_co2_reduced:585.09',
    'temperature:46.8',
    'state:Normal',
    'today_max_current:2930',
  ])
# ---
# name: test_eth[status2.xml]
  list([
    'current_power:4986',
    'grid_voltage:243.0',
    'grid_l1:None',
    'grid_l2:None',
    'grid_l3:None',
    'today_yield:14.7',
    'total_yield:2340.2',
    'today_time:8.5',
    'total_time:2267.4',
    'total_co2_reduced:2333.18',
    'temperature:46.8',
    'state:Normal',
    'today_max_current:None',
  ])
# ---
# name: test_eth[status3.xml]
  list([
    'current_power:449',
    'grid_voltage:227.2',
    'grid_l1:None',
    'grid_l2:None',
    'grid_l3:None',
    'today_yield:3.2',
    'total_yield:11356.8',
    'today_time:5.6',
    'total_time:18549.1',
    'total_co2_reduced:11322.73',
    'temperature:29.8',
    'state:Normal',
    'today_max_current:None',
  ])
# ---
# name: test_wifi[0-15020J2012EN02905]
  list([
    'current_power:1921',
    'grid_voltage:None',
    'grid_l1:244.7',
    'grid_l2:None',
    'grid_l3:None',
    'today_yield:19.81',
    'total_yield:12545.79',
    'today_time:11.3',
    'total_time:11652.8',
    'total_co2_reduced:9848.4',
    'temperature:45.3',
    'state:Normal',
    'today_max_current:None',
  ])
# ---
# name: test_wifi[1-21030G1628XXYYYYY]
  list([
    'current_power:1040',
    'grid_voltage:None',
    'grid_l1:241.3',
    'grid_l2:None',
    'grid_l3:None',
    'today_yield:21.07',
    'total_yield:12547.05',
    'today_time:12.5',
    'total_time:11654.0',
    'total_co2_reduced:9849.4',
    'temperature:41.5',
    'state:Normal',
    'today_max_current:None',
  ])
# ---
